notebooks/generative/fashionmnist_autoencoder.livemd

```diff
@@ -73,142 +63,150 @@ Next, we pass the output of the encoder to the decoder and try to reconstruct th
+Finally, we can train the model. We'll use the `:adam` and `:mean_squared_error` loss with `Axon.Loop.trainer`. Our loss function will measure the aggregate error between pixels of original images and the model's reconstructed images. We'll also `:mean_absolute_error` using `Axon.Loop.metric`. `Axon.Loop.run` trains the model with the given training data.
 
 ```elixir
-defmodule Evaluation do
-  require Axon
-
-  def run do
-    {images, _} = DataSource.get_data()
+batch_size = 32
+epochs = 5
 
-    train_images = Preprocessing.transform_images(images)
+batched_images = Nx.to_batched(train_images, batch_size)
+train_batches = Stream.zip(batched_images, batched_images)
 
-    model = Autoencoder.build_model({nil, 1, 28, 28}, 64) |> IO.inspect()
+params =
+  model
+  |> Axon.Loop.trainer(:mean_squared_error, :adam)
+  |> Axon.Loop.metric(:mean_absolute_error, "Error")
+  |> Axon.Loop.run(train_batches, %{}, epochs: epochs, compiler: EXLA)
+```
 
-    model_state = Trainer.train_model(model, train_images, 5)
+## Extra: losses
 
-    sample_image =
-      train_images
-      |> hd()
-      |> Nx.slice_axis(0, 1, 0)
-      |> Nx.reshape({1, 1, 28, 28})
+To better understand what is mean absolute error (MAE) and mean square error (MSE) let's go through an example.
 
-    sample_image |> Nx.to_heatmap() |> IO.inspect()
+```elixir
+mean_square_error = fn y_pred, y ->
+  y_pred
+  |> Nx.subtract(y)
+  |> Nx.power(2)
+  |> Nx.mean(axes: [-1])
+  |> Nx.sum()
+end
 
-    model
-    |> Axon.predict(model_state, sample_image, compiler: EXLA)
-    |> Nx.to_heatmap()
-    |> IO.inspect()
-  end
+mean_absolute_erorr = fn y_pred, y ->
+  y_pred
+  |> Nx.subtract(y)
+  |> Nx.abs()
+  |> Nx.mean(axes: [-1])
+  |> Nx.sum()
 end
 ```
 
-## Results analysis
+We will work with a sample image of a shoe, a slightly noised version of that image, and also an entirely different image from the dataset.
+
+```elixir
+shoe_image = train_images[0..0//1]
+noised_shoe_image = Nx.add(shoe_image, Nx.random_normal(shoe_image, 0.0, 0.05))
+other_image = train_images[1..1//1]
+:ok
+```
+
+For the same image both errors should be 0, because when we have two exact copies, there is no pixel difference.
+
+```elixir
+{
+  mean_square_error.(shoe_image, shoe_image),
+  mean_absolute_erorr.(shoe_image, shoe_image)
+}
+```
+
+Now the noised image:
+
+```elixir
+{
+  mean_square_error.(shoe_image, noised_shoe_image),
+  mean_absolute_erorr.(shoe_image, noised_shoe_image)
+}
+```
+
+And a different image:
+
+```elixir
+{
+  mean_square_error.(shoe_image, other_image),
+  mean_absolute_erorr.(shoe_image, other_image)
+}
+```
+
+As we can see, the noised image has a non-zero MSE and MAE but is much smaller than the error of two completely different pictures. In other words, both of these error types measure the level of similarity between images. A small error implies decent prediction values. On the other hand, a large error value suggests poor quality of predictions.
```

Comment:
You don't have to include this, just a fun fact. MAE and MSE can also be called the L1 and L2 loss respectively for the L1 and L2 norm. The L2 loss (MSE) is typically preferred because it's a smoother function whereas L1 is often difficult to optimize with SGD.

Reply:
Worth including, thanks! I also changed the MAE and MSE definitions, because I think they were not adjusted for the multidimensional images. I changed them to work on a single example for simplicity. Please check if it makes sense (4aa162096e4562b11fc280282bc90ca98c1840cb) and feel free to merge then :)